Microsoft.Toolkit.Uwp.UI/Extensions/Markup/OnDevice.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Windows.ApplicationModel;
+using Windows.UI.Xaml.Markup;
+
+namespace Microsoft.Toolkit.Uwp.UI.Extensions.Markup
+{
+    /// <summary>
+    /// The OnDevice markup extension allows you to customize UI appearance on a per-DeviceFamily basis.
+    /// </summary>
+    /// <example>
+    /// The next TextBlock will show the text 'Hello' on desktop computers, 'World' on Xbox and 'Hi' on all other devices
+    /// <code>
+    ///     <TextBlock Text="{helpers:OnDevice Default=Hi, Desktop=Hello, Xbox=World}"
+    ///                xmlns:helpers="using:Microsoft.Toolkit.Uwp.UI.Extensions.Markup" />
+    /// </code>
+    /// </example>
+    public class OnDevice : MarkupExtension
+    {
+        /// <summary>
+        /// Gets or sets the default value for this property
+        /// </summary>
+        public object Default { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value for this property on Desktop
+        /// </summary>
+        public object Desktop { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value for this property on Holographic
+        /// </summary>
+        public object Holographic { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value for this property on IoT
+        /// </summary>
+        public object IoT { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value for this property on Team
+        /// </summary>
+        public object Team { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value for this property on Xbox
+        /// </summary>
+        public object Xbox { get; set; }
+
+        /// <summary>
+        /// Returns an object value for the current DeviceFamily, Default is used when it is not set.
+        /// </summary>
+        /// <returns>The object value to set on the property where the extension is applied.</returns>
+        protected override object ProvideValue()
+        {
+            string deviceFamily = null;
+            if (DesignMode.DesignMode2Enabled)
+            {
+                // TODO: detect DeviceFamily in XAML Designer (device dropdown)
+                // deviceFamily = ???
+            } else
+            {
+                deviceFamily = Windows.System.Profile.AnalyticsInfo.VersionInfo.DeviceFamily;
+            }
+
+            switch (deviceFamily)
+            {
+                case "Windows.Desktop":
+                    return this.Desktop ?? this.Default;
+                case "Windows.Holographic":
+                    return this.Holographic ?? this.Default;
+                case "Windows.IoT":
```

Comment:
I noticed our [AnalyticsExtension](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/master/Microsoft.Toolkit.Uwp.SampleApp/Common/AnalyticsVersionInfoExtensions.cs#L40) have this as `Windows.Universal` which is wrong in the SampleApp.  It could however also be `IoTHeadless` but that wouldn't make sense for a XAML app.

I think we should still move the AnalyticsVersionInfoExtensions into the `Microsoft.Toolkit.Uwp` namespace to centralize this code for here and the sample app.  Then you can just look at the enum values and condense the cases down to the ones the markup extension cares about (e.g. Desktop and Server all go to your Desktop one).  That would mean we drop the `Tablet` value anyway, as we don't want to pull in a UI dependency there, plus as you mentioned that value can change dynamically (though I am curious if we could return a binding from a markup extension that could be mutable...).

```
    public enum DeviceFormFactor
    {
        Other,
        Desktop,
        Mobile,
        Xbox,
        Team,
        IoT,
        IoTHeadless,
        Server,
        Holographic,
    }
```

Reply:
It is absolutely interesting if we could return a binding. I have never tried that.

Reply:
Mobile doesn't make sense either because the toolkit (+MarkupExtensions) requires SDK 16299 which is not supported on Mobile.

Reply:
Correct, Mobile doesn't make sense for this extension, but would for the enum returned in a centralized extension to AnalyticsVersionInfo.

Reply:
Is this a blocking issue for this PR?

Reply:
No, we can do it later if it's needed.  So, think we're good.  Thanks for the discussion!